src/System.Data.Common/src/System/Data/Common/DbProviderFactories.cs

```diff
@@ -0,0 +1,187 @@
+                    }
+                    throw ADP.Argument(SR.Format(SR.ADP_DbProviderFactories_FactoryNotLoadable, assemblyQualifiedName));
+                }
+            }
+            throw ADP.Argument(SR.ADP_DbProviderFactories_NoAssemblyQualifiedName);
+        }
+
+        public static DataTable GetFactoryClasses()
+        {
+            try
+            {
+                _providerTableLock.EnterReadLock();
+                return _providerTable == null ? GetInitialProviderTable() : _providerTable.Copy();
+            }
+            finally
+            {
+                _providerTableLock.ExitReadLock();
+            }
+        }
+
+        public static DbProviderFactory GetFactory(DbConnection connection)
+        {
+            ADP.CheckArgumentNull(connection, nameof(connection));
+            return connection.ProviderFactory;
+        }
+
+        public static void ConfigureFactory(Type providerFactoryClass, string providerInvariantName)
+        {
+            ConfigureFactory(providerFactoryClass, providerInvariantName, string.Empty, string.Empty);
+        }
+        
+        public static void ConfigureFactory(Type providerFactoryClass, string providerInvariantName, string name, string description)
+        {
+            ADP.CheckArgumentNull(providerFactoryClass, nameof(providerFactoryClass));
+            ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+
+            if (!providerFactoryClass.IsSubclassOf(typeof(DbProviderFactory)))
+            {
+                throw ADP.Argument(SR.Format(SR.ADP_DbProviderFactories_NotAFactoryType, providerFactoryClass.FullName));
+            }
+            RegisterFactoryInTable(providerFactoryClass, providerInvariantName, name, description);
+        }
+        
+        public static void ConfigureFactory(DbConnection connection, string providerInvariantName)
+        {
+            ConfigureFactory(connection, providerInvariantName, string.Empty, string.Empty);
+        }
+        
+        public static void ConfigureFactory(DbConnection connection, string providerInvariantName, string name, string description)
+        {
+            ADP.CheckArgumentNull(connection, nameof(connection));
+            ADP.CheckArgumentLength(providerInvariantName, nameof(providerInvariantName));
+
+            DbProviderFactory factoryInstance = GetFactory(connection);
+            if (factoryInstance == null)
+            {
+                throw ADP.Argument(SR.ADP_DbProviderFactories_NoFactorySuppliedThroughDbConnection);
+            }
+            RegisterFactoryInTable(factoryInstance.GetType(), providerInvariantName, name, description);
+        }
+        
+        private static void RegisterFactoryInTable(Type factoryType, string providerInvariantName, string name, string description)
+        {
+            ADP.CheckArgumentNull(factoryType, nameof(factoryType));
```

Comment:
Looking at the ADO.Net code, it uses Debug.Assert over multiple checks in the same code path. It would be good to avoid many checks in the same code path for the same condition. 
No `var` please.